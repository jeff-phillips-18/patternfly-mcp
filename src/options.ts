import { join } from 'node:path';
import packageJson from '../package.json';

/**
 * CLI options that users can set via command line arguments
 */
interface CliOptions {
  docsHost?: boolean;
  // Future CLI options can be added here
}

/**
 * Application defaults (not user-configurable)
 */
interface AppDefaults {
  resourceMemoOptions: typeof RESOURCE_MEMO_OPTIONS;
  toolMemoOptions: typeof TOOL_MEMO_OPTIONS;
  pfExternal: string;
  pfExternalCharts: string;
  pfExternalChartsComponents: string;
  pfExternalChartsDesign: string;
  pfExternalDesign: string;
  pfExternalDesignComponents: string;
  pfExternalDesignLayouts: string;
  pfExternalAccessibility: string;
  separator: string;
  urlRegex: RegExp;
  name: string;
  version: string;
  repoName: string|undefined;
  contextPath: string;
  docsPath: string;
  llmsFilesPath: string;
}

/**
 * Frozen options object (immutable configuration)
 */
interface GlobalOptions extends CliOptions, AppDefaults {
  // This will be frozen and immutable
}

/**
 * Default separator for joining multiple document contents
 */
const DEFAULT_SEPARATOR = '\n\n---\n\n';

/**
 * Resource-level memoization options
 */
const RESOURCE_MEMO_OPTIONS = {
  fetchUrl: {
    cacheLimit: 100,
    expire: 3 * 60 * 1000, // 3 minute sliding cache
    cacheErrors: false
  },
  readFile: {
    cacheLimit: 50,
    expire: 2 * 60 * 1000, // 2 minute sliding cache
    cacheErrors: false
  }
};

/**
 * Tool-specific memoization options
 */
const TOOL_MEMO_OPTIONS = {
  usePatternFlyDocs: {
    cacheLimit: 10,
    expire: 1 * 60 * 1000, // 1 minute sliding cache
    cacheErrors: false
  },
  fetchDocs: {
    cacheLimit: 15,
    expire: 1 * 60 * 1000, // 1 minute sliding cache
    cacheErrors: false
  }
};

/**
 * URL regex pattern for detecting external URLs
 */
const URL_REGEX = /^(https?:)\/\//i;

/**
 * PatternFly docs root URL
 */
const PF_EXTERNAL = 'https://raw.githubusercontent.com/patternfly/patternfly-org/refs/heads/main/packages/documentation-site/patternfly-docs/content';

/**
 * PatternFly design guidelines URL
 */
const PF_EXTERNAL_DESIGN = `${PF_EXTERNAL}/design-guidelines`;

/**
 * PatternFly design guidelines' components' URL
 */
const PF_EXTERNAL_DESIGN_COMPONENTS = `${PF_EXTERNAL_DESIGN}/components`;

/**
 * PatternFly design guidelines' layouts' URL
 */
const PF_EXTERNAL_DESIGN_LAYOUTS = `${PF_EXTERNAL_DESIGN}/layouts`;

/**
 * PatternFly accessibility URL
 */
const PF_EXTERNAL_ACCESSIBILITY = `${PF_EXTERNAL}/accessibility`;

/**
 * PatternFly charts root URL
 */
const PF_EXTERNAL_CHARTS = 'https://raw.githubusercontent.com/patternfly/patternfly-react/refs/heads/main/packages/react-charts/src';

/**
 * PatternFly charts' components' URL
 */
const PF_EXTERNAL_CHARTS_COMPONENTS = `${PF_EXTERNAL_CHARTS}/victory/components`;

/**
 * PatternFly charts' design guidelines URL
 */
const PF_EXTERNAL_CHARTS_DESIGN = `${PF_EXTERNAL_CHARTS}/charts`;

/**
 * Global configuration options object.
 *
 * @type {GlobalOptions}
 * @property {CliOptions.docsHost} [docsHost] - Flag indicating whether to use the docs-host.
 * @property {string} pfExternal - PatternFly external docs URL.
 * @property {string} pfExternalCharts - PatternFly external charts URL.
 * @property {string} pfExternalChartsComponents - PatternFly external charts components URL.
 * @property {string} pfExternalChartsDesign - PatternFly external charts design guidelines URL.
 * @property {string} pfExternalDesign - PatternFly external design guidelines URL.
 * @property {string} pfExternalDesignComponents - PatternFly external design guidelines components URL.
 * @property {string} pfExternalDesignLayouts - PatternFly external design guidelines layouts URL.
 * @property {string} pfExternalAccessibility - PatternFly external accessibility URL.
 * @property {typeof RESOURCE_MEMO_OPTIONS} resourceMemoOptions - Resource-level memoization options.
 * @property {typeof TOOL_MEMO_OPTIONS} toolMemoOptions - Tool-specific memoization options.
 * @property {string} separator - Default string delimiter.
 * @property {RegExp} urlRegex - Regular expression pattern for URL matching.
 * @property {string} name - Name of the package.
 * @property {string} version - Version of the package.
 * @property {string} repoName - Name of the repository.
 * @property {string} contextPath - Current working directory.
 * @property {string} docsPath - Path to the documentation directory.
 * @property {string} llmsFilesPath - Path to the LLMs files directory.
 */
const OPTIONS: GlobalOptions = {
  pfExternal: PF_EXTERNAL,
  pfExternalCharts: PF_EXTERNAL_CHARTS,
  pfExternalChartsComponents: PF_EXTERNAL_CHARTS_COMPONENTS,
  pfExternalChartsDesign: PF_EXTERNAL_CHARTS_DESIGN,
  pfExternalDesign: PF_EXTERNAL_DESIGN,
  pfExternalDesignComponents: PF_EXTERNAL_DESIGN_COMPONENTS,
  pfExternalDesignLayouts: PF_EXTERNAL_DESIGN_LAYOUTS,
  pfExternalAccessibility: PF_EXTERNAL_ACCESSIBILITY,
  resourceMemoOptions: RESOURCE_MEMO_OPTIONS,
  toolMemoOptions: TOOL_MEMO_OPTIONS,
  separator: DEFAULT_SEPARATOR,
  urlRegex: URL_REGEX,
  name: packageJson.name,
  version: packageJson.version,
  repoName: process.cwd()?.split?.('/')?.pop?.()?.trim?.(),
  contextPath: (process.env.NODE_ENV === 'local' && '/') || process.cwd(),
  docsPath: (process.env.NODE_ENV === 'local' && '/documentation') || join(process.cwd(), 'documentation'),
  llmsFilesPath: (process.env.NODE_ENV === 'local' && '/llms-files') || join(process.cwd(), 'llms-files'),
};

/**
 * Parse CLI arguments and return CLI options
 */
const parseCliOptions = (): CliOptions => {
  return {
    docsHost: process.argv.includes('--docs-host'),
    // Future CLI options can be added here
  };
};

/**
 * Make global options immutable after combining CLI options with app defaults.
 */
const freezeOptions = (cliOptions: CliOptions) => {
  Object.assign(OPTIONS, {
    ...cliOptions
  });

  return Object.freeze(OPTIONS);
};

export {
  parseCliOptions,
  freezeOptions,
  OPTIONS,
  PF_EXTERNAL,
  PF_EXTERNAL_CHARTS,
  PF_EXTERNAL_CHARTS_COMPONENTS,
  PF_EXTERNAL_CHARTS_DESIGN,
  PF_EXTERNAL_DESIGN,
  PF_EXTERNAL_DESIGN_COMPONENTS,
  PF_EXTERNAL_DESIGN_LAYOUTS,
  PF_EXTERNAL_ACCESSIBILITY,
  RESOURCE_MEMO_OPTIONS,
  TOOL_MEMO_OPTIONS,
  DEFAULT_SEPARATOR,
  URL_REGEX,
  type CliOptions,
  type AppDefaults,
  type GlobalOptions
};
