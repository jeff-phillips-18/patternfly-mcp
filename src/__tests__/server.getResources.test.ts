import { readFile } from 'node:fs/promises';
import { readLocalFileFunction, fetchUrlFunction, resolveLocalPathFunction, processDocsFunction } from '../server.getResources';
import { type GlobalOptions } from '../options';

// Mock dependencies
jest.mock('node:fs/promises');
jest.mock('../server.caching', () => ({
  memo: jest.fn(fn => {
    const memoized = fn;

    memoized.clear = jest.fn();

    return memoized;
  })
}));

const mockReadFile = readFile as jest.MockedFunction<typeof readFile>;

describe('readLocalFileFunction', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt to read a file from disk', async () => {
    mockReadFile.mockResolvedValue('file content');

    const result = await readLocalFileFunction('/path/to/file.md');

    expect(mockReadFile).toHaveBeenCalledWith('/path/to/file.md', 'utf-8');
    expect(result).toBe('file content');
  });

  it('should have memo property', () => {
    expect(readLocalFileFunction.memo).toBeDefined();
  });
});

describe('fetchUrlFunction', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    global.fetch = jest.fn();
  });

  it('should attempt to fetch a URL with correct headers', async () => {
    const mockResponse = {
      ok: true,
      text: jest.fn().mockResolvedValue('fetched content')
    };

    (global.fetch as jest.Mock).mockResolvedValue(mockResponse);

    const result = await fetchUrlFunction('https://example.com/doc.md');

    expect(global.fetch).toHaveBeenCalledWith(
      'https://example.com/doc.md',
      expect.objectContaining({
        headers: { Accept: 'text/plain, text/markdown, */*' }
      })
    );
    expect(result).toBe('fetched content');
  });

  it('should handle non-OK responses', async () => {
    const mockResponse = {
      ok: false,
      status: 404,
      statusText: 'Not Found'
    };

    (global.fetch as jest.Mock).mockResolvedValue(mockResponse);

    await expect(fetchUrlFunction('https://example.com/missing.md'))
      .rejects
      .toThrow('Failed to fetch https://example.com/missing.md: 404 Not Found');
  });

  it('should have memo property', () => {
    expect(fetchUrlFunction.memo).toBeDefined();
  });
});

describe('resolveLocalPathFunction', () => {
  it.each([
    {
      description: 'with docsHost true',
      options: {
        docsHost: true,
        llmsFilesPath: '/llms-files'
      },
      path: 'react-core/6.0.0/llms.txt'
    },
    {
      description: 'with docsHost false',
      options: {
        docsHost: false,
        llmsFilesPath: '/llms-files'
      },
      path: 'documentation/README.md'
    }
  ])('should return a consistent path, $description', ({ path, options }) => {
    const result = resolveLocalPathFunction(path, options as GlobalOptions);

    expect(result).toMatchSnapshot();
  });
});

describe('processDocsFunction', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    // Mock the memo functions
    readLocalFileFunction.memo = jest.fn().mockResolvedValue('local file content');
    fetchUrlFunction.memo = jest.fn().mockResolvedValue('fetched content');
  });

  it.each([
    {
      description: 'files and URLS',
      inputs: [
        'local-file.md',
        'https://example.com/remote.md'
      ],
      options: {
        docsHost: false,
        urlRegex: /^(https?:)\/\//i,
        separator: '\n\n---\n\n',
        llmsFilesPath: '/llms-files'
      },
      fileMemoHits: 1,
      fetchMemoHits: 1
    },
    {
      description: 'duplicate files and URLS',
      inputs: [
        'file.md',
        'file.md',
        'file.md',
        'https://example.com/remote.md',
        'https://example.com/remote.md'
      ],
      options: {
        docsHost: false,
        urlRegex: /^(https?:)\/\//i,
        separator: '\n\n---\n\n',
        llmsFilesPath: '/llms-files'
      },
      fileMemoHits: 1,
      fetchMemoHits: 1
    },
    {
      description: 'filter empty strings',
      inputs: [
        'file.md',
        '',
        '   ',
        'file2.md'
      ],
      options: {
        docsHost: false,
        urlRegex: /^(https?:)\/\//i,
        separator: '\n\n---\n\n',
        llmsFilesPath: '/llms-files'
      },
      fileMemoHits: 2
    }
  ])('should process local and remote inputs, $description', async ({ inputs, options, fileMemoHits = 0, fetchMemoHits = 0 }) => {
    const result = await processDocsFunction(inputs, options as GlobalOptions);

    expect(result).toMatchSnapshot();
    expect(readLocalFileFunction.memo).toHaveBeenCalledTimes(fileMemoHits);
    expect(fetchUrlFunction.memo).toHaveBeenCalledTimes(fetchMemoHits);
  });

  it('should handle errors gracefully', async () => {
    const mockOptions = {
      docsHost: false,
      urlRegex: /^(https?:)\/\//i,
      separator: '\n\n---\n\n',
      llmsFilesPath: '/llms-files'
    };

    // Mock one success and one failure
    readLocalFileFunction.memo = jest.fn()
      .mockResolvedValueOnce('success content')
      .mockRejectedValueOnce(new Error('File not found'));

    const inputs = [
      'good-file.md',
      'bad-file.md'
    ];

    const result = await processDocsFunction(inputs, mockOptions as GlobalOptions);

    expect(result).toMatchSnapshot('errors');
  });
});
