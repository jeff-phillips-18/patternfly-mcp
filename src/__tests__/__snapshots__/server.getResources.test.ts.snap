// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`processDocsFunction should handle errors gracefully: errors 1`] = `
"# Documentation from good-file.md

success content

---

‚ùå Failed to load bad-file.md: Error: File not found"
`;

exports[`processDocsFunction should process local and remote inputs, duplicate files and URLS 1`] = `
"# Documentation from file.md

local file content

---

# Documentation from https://example.com/remote.md

fetched content"
`;

exports[`processDocsFunction should process local and remote inputs, files and URLS 1`] = `
"# Documentation from local-file.md

local file content

---

# Documentation from https://example.com/remote.md

fetched content"
`;

exports[`processDocsFunction should process local and remote inputs, filter empty strings 1`] = `
"# Documentation from file.md

local file content

---

# Documentation from file2.md

local file content"
`;

exports[`resolveLocalPathFunction should return a consistent path, with docsHost false 1`] = `"documentation/README.md"`;

exports[`resolveLocalPathFunction should return a consistent path, with docsHost true 1`] = `"/llms-files/react-core/6.0.0/llms.txt"`;
